cache.js

```diff
@@ -215,162 +355,232 @@ function serverSide(cache, app, options)
                     {
                         modelName: modelName,
                         methodName: 'delete',
                         modelId: curr.id
-                    }).catch(console.error);
+                    }, topicName).catch(console.error);
                 });
             }, Promise.resolve());
         }).catch(console.error).then(next);
     }
+}
 
-    cache.emit = function (data)
+
+/* Cache helpers*/
+
+//Query db for requested data
+function getCacheData(app, cache, data)
+{
+    const dbProms = [];
+    for (let modelName in data)
     {
-        return cache.receivers.reduce(function (prev, curr, idx)
+        //Apply publishing hooks to relevant models
+        const Model = app.models[modelName];
+        if (cache.modelsWatched.indexOf(modelName) < 0) setModelsWatched(app, cache, [modelName]);
+
+        if (data[modelName].type && data[modelName].type !== 'cache') continue;
+
+        //Only prime the cache if the type is cache or left blank
+        const query = {};
+        if (data[modelName].fields && data[modelName].fields.length) query.fields = data[modelName].fields;
+        const prom = Model.find(query).then(modelData =>
         {
-            return prev.then(function ()
-            {
-                return cache.send(curr, data).catch(console.error);
-            });
-        }, Promise.resolve());
+            return {
+                modelName: modelName,
+                data: modelData
+            };
+        });
+        dbProms.push(prom);
     }
+    return Promise.all(dbProms);
 }
 
+//Apply hook handlers to watched models
+function setModelsWatched(app, cache, models)
+{
+    if (!models || !models.length) return;
+    models.forEach(m =>
+    {
+        const Model = app.models[m];
+        if (!Model) return;
+
+        cache.modelsWatched.push(m);
+
+        Model.observe('after save', afterSaveHook(cache));
+        Model.observe('before delete', beforeDeleteHook(cache));
+    });
+}
+
+
+/* General helpers */
+
+function getModelName(ctx)
+{
+    return ctx.Model && ctx.Model.definition && ctx.Model.definition.name;
+}
+
+function getType(val)
+{
+    return Object.prototype.toString.call(val).slice(8, -1);
+}
+
+
+
+/* Cache Machine types */
+
 /**
- * Local side broadcasting
+ * Client side listening
  */
-function localSide(cache, app, options)
+function clientSide(cache, options)
 {
-    if (!app) throw new Error('app is required for cache server');
+    if (!options.projectId) throw new Error('Google Project Id is required for cache client');
+    if (options.eventConfig && options.eventConfig.events && !options.eventConfig.eventFn) throw new Error('options.eventConfig.eventFn is required if including events');
 
-    cache.modelsWatched = [];
+    let modelsToNotify = {};
 
-    cache.ask = function (modelName)
-    {
-        var Model = app.models[modelName];
+    cache.modelsToWatch = {};
+    cache.watchModels(cache.modelsToWatch, (options.modelsToWatch || []));
```

Comment:
would there be a nicer way to do this instead of instanciating cache.modelsToWatch = {} and then passing that? Instead do something like `cache.modelsToWatch = cache.getModelsToWatch(options.modelsToWatch || []);` ?

Reply:
Yeah, I think this is a relic of an older version where I was thinking about local version passing in a different object, but that ended up not happening.